File: leetcode/python/Sort List.py
# Sort List
# for leetcode problems
# 2014.08.13 by zhanglin

# Problem:
# Sort a linked list in O(n log n) time using constant space complexity.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Divide and Conquer solution
class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def merge(self, lst1, lst2):
        if lst1 == None:
            return lst2
        if lst2 == None:
            return lst1

        new_lst = ListNode(0) # new a linked list
        pointer = new_lst

        while lst1 and lst2: # don't use lst1.next != None judgment, that will lead TLE problem
            if lst1.val <= lst2.val:
                pointer.next = lst1
                lst1 = lst1.next
            else:
                pointer.next = lst2
                lst2 = lst2.next
            pointer = pointer.next # point to next

        if lst1 == None: # if one list is out, add another to the end
            pointer.next = lst2
        if lst2 == None:
            pointer.next = lst1

        return new_lst.next

        
    def sortList(self, head):
        if head == None or head.next == None: # only one node
            return head

        fast = head
        slow = head

        while fast.next and fast.next.next: # don't use fast.next != None judgment, that will lead TLE problem
            fast = fast.next.next
            slow = slow.next

        new_head_1st = head
        new_head_2nd = slow.next
        slow.next = None # cut the linked list

        new_head_1st = self.sortList(new_head_1st)
        new_head_2nd = self.sortList(new_head_2nd)

        return self.merge(new_head_1st, new_head_2nd)

File: leetcode/c++/Sort List.cpp
// Sort List
// for leetcode problems
// 2014.08.14 by zhanglin

// Problem:
// Sort a linked list in O(n log n) time using constant space complexity.

/**
 * Definition for singly-linked list->
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *merge(ListNode *left, ListNode *right)
    {
        ListNode *new_head;
        ListNode *pointer;
        
        if(left == NULL)
        {
            return right;
        }
        if(right == NULL)
        {
            return left;
        }
        
        new_head = new ListNode(0);
        pointer = new_head;
        
        while(left != NULL && right != NULL)
        {
            if(left->val < right->val)
            {
                pointer->next = left;
                left = left->next;
            }
            else
            {
                pointer->next = right;
                right = right->next;
            }
            pointer = pointer->next;
        }
        
        if(left == NULL)
        {
            pointer->next = right;
        }
        if(right == NULL)
        {
            pointer->next = left;
        }
        
        return new_head->next;
    }
    
    ListNode *sortList(ListNode *head)
    {
        ListNode *fast;
        ListNode *slow;
        ListNode *new_head_left;
        ListNode *new_head_right;
        
        if(head == NULL || head->next == NULL)
        {
            return head;
        }
        
        fast = head;
        slow = head;
        
        while(fast->next != NULL && fast->next->next != NULL)
        {
            fast = fast->next->next;
            slow = slow->next;
        }
        
        new_head_left = head;
        new_head_right = slow->next;
        slow->next = NULL;
        
        new_head_left = sortList(new_head_left);
        new_head_right = sortList(new_head_right);
        
        return merge(new_head_left, new_head_right);
    }
};


